File: backend/app/web/paymongo.py
"""
PayMongo Payment Integration Module
Handles e-wallet and card payment processing with PayMongo API
"""
import os
import requests
import base64
from typing import Dict, Optional, Any
from dotenv import load_dotenv

load_dotenv()

# PayMongo API Configuration
PAYMONGO_SECRET_KEY = os.getenv("PAYMONGO_SECRET_KEY", "sk_test_")
PAYMONGO_PUBLIC_KEY = os.getenv("PAYMONGO_PUBLIC_KEY", "pk_test_")
PAYMONGO_TEST_MODE = True  # Always use test mode by default

# API Endpoints
PAYMONGO_API_URL = "https://api.paymongo.com/v1"

def _get_auth_header():
    """Generate Basic Auth header for PayMongo API"""
    # Use secret key for Authorization header
    credentials = base64.b64encode(f"{PAYMONGO_SECRET_KEY}:".encode()).decode()
    return {"Authorization": f"Basic {credentials}"}


def create_payment_intent_for_ewallet(
    amount: int,  # in centavos (PHP)
    provider: str,  # 'gcash', 'grabpay', 'maya'
    description: str = "",
    redirect_url: str = ""
) -> Dict[str, Any]:
    """
    Create a payment intent for e-wallet (GCash/GrabPay/Maya)
    Uses redirect flow - returns checkout_url for user to complete payment
    
    Args:
        amount: Amount in PHP centavos (e.g., 50000 = PHP 500.00)
        provider: E-wallet provider ('gcash', 'grabpay', 'maya')
        description: Order description
        redirect_url: URL to redirect after payment completion
    
    Returns:
        Dict with checkout_url and payment_intent_id on success
        Dict with error details on failure
    """
    try:
        # Map provider names to PayMongo format
        provider_map = {
            'gcash': 'gcash',
            'grabpay': 'grab_pay',
            'maya': 'paymaya'
        }
        
        paymongo_provider = provider_map.get(provider.lower(), provider)
        
        payload = {
            "data": {
                "attributes": {
                    "amount": amount,  # in centavos
                    "currency": "PHP",
                    "description": description,
                    "statement_descriptor": "DaingGrader",
                    "payment_method_allowed": [paymongo_provider],
                    "redirect": {
                        "success": redirect_url,
                        "failed": redirect_url,
                        "cancelled": redirect_url,
                    },
                }
            }
        }
        
        response = requests.post(
            f"{PAYMONGO_API_URL}/payment_intents",
            json=payload,
            headers=_get_auth_header(),
            timeout=10
        )
        
        if response.status_code not in [200, 201]:
            return {
                "success": False,
                "error": f"PayMongo API error: {response.status_code}",
                "details": response.text
            }
        
        data = response.json().get("data", {})
        payment_intent = data.get("attributes", {})
        
        # Get the checkout URL
        client_key = data.get("id", "")
        checkout_url = payment_intent.get("checkout_url", "")
        
        if not checkout_url:
            return {
                "success": False,
                "error": "No checkout URL returned from PayMongo"
            }
        
        return {
            "success": True,
            "payment_intent_id": client_key,
            "checkout_url": checkout_url,
            "amount": amount,
            "provider": provider
        }
        
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


def create_payment_intent_for_card(
    amount: int,
    description: str = "",
    redirect_url: str = ""
) -> Dict[str, Any]:
    """
    Create a payment intent for card payment
    
    Args:
        amount: Amount in PHP centavos
        description: Order description
        redirect_url: URL to redirect after payment
    
    Returns:
        Dict with payment_intent_id and client_key
    """
    try:
        payload = {
            "data": {
                "attributes": {
                    "amount": amount,
                    "currency": "PHP",
                    "description": description,
                    "statement_descriptor": "DaingGrader",
                    "payment_method_allowed": ["card"],
                    "redirect": {
                        "success": redirect_url,
                        "failed": redirect_url,
                        "cancelled": redirect_url,
                    },
                }
            }
        }
        
        response = requests.post(
            f"{PAYMONGO_API_URL}/payment_intents",
            json=payload,
            headers=_get_auth_header(),
            timeout=10
        )
        
        if response.status_code not in [200, 201]:
            return {
                "success": False,
                "error": f"PayMongo API error: {response.status_code}"
            }
        
        data = response.json().get("data", {})
        client_key = data.get("id", "")
        
        return {
            "success": True,
            "payment_intent_id": client_key,
            "client_key": client_key
        }
        
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


def retrieve_payment_intent(payment_intent_id: str) -> Dict[str, Any]:
    """
    Retrieve payment intent details from PayMongo
    Used to check payment status
    
    Args:
        payment_intent_id: PayMongo payment intent ID
    
    Returns:
        Payment intent data including status
    """
    try:
        response = requests.get(
            f"{PAYMONGO_API_URL}/payment_intents/{payment_intent_id}",
            headers=_get_auth_header(),
            timeout=10
        )
        
        if response.status_code != 200:
            return {
                "success": False,
                "error": f"Failed to retrieve payment intent: {response.status_code}"
            }
        
        data = response.json().get("data", {})
        attributes = data.get("attributes", {})
        
        return {
            "success": True,
            "id": data.get("id"),
            "status": attributes.get("status"),  # succeeded, failed, etc
            "amount": attributes.get("amount"),
            "currency": attributes.get("currency"),
            "description": attributes.get("description"),
            "payment_method": attributes.get("payments", [{}])[0] if attributes.get("payments") else None
        }
        
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


def attach_payment_method_to_intent(
    payment_intent_id: str,
    payment_method_id: str
) -> Dict[str, Any]:
    """
    Attach a payment method to a payment intent (for card payments)
    
    Args:
        payment_intent_id: PayMongo payment intent ID
        payment_method_id: PayMongo payment method ID (from tokenization)
    
    Returns:
        Result of the attachment
    """
    try:
        payload = {
            "data": {
                "attributes": {
                    "payment_method": payment_method_id
                }
            }
        }
        
        response = requests.post(
            f"{PAYMONGO_API_URL}/payment_intents/{payment_intent_id}/attach",
            json=payload,
            headers=_get_auth_header(),
            timeout=10
        )
        
        if response.status_code not in [200, 201]:
            return {
                "success": False,
                "error": f"Failed to attach payment method: {response.status_code}"
            }
        
        return {"success": True}
        
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }

File: daing-grader-web/src/services/paymongo.ts
/**
 * PayMongo API Service
 * Handles payment processing with PayMongo
 */

import { PAYMONGO_CONFIG } from '../config/paymongo'

export interface CardDetails {
  number: string
  expMonth: string
  expYear: string
  cvc: string
}

export interface PayMongoPaymentMethod {
  id: string
  type: string
  attributes: {
    type: string
    details: any
  }
}

export interface PayMongoPaymentIntent {
  id: string
  attributes: {
    amount: number
    currency: string
    status: string
    payment_method_allowed: string[]
  }
}

/**
 * Create a PayMongo Payment Method (Tokenize card)
 */
export async function createPaymentMethod(cardDetails: CardDetails): Promise<PayMongoPaymentMethod> {
  const { number, expMonth, expYear, cvc } = cardDetails

  const payload = {
    data: {
      attributes: {
        type: 'card',
        details: {
          card_number: number.replace(/\s/g, ''),
          exp_month: parseInt(expMonth),
          exp_year: parseInt(expYear),
          cvc: cvc,
        },
      },
    },
  }

  try {
    const response = await fetch(`${PAYMONGO_CONFIG.paymongoApiUrl}/payment_methods`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Basic ${btoa(PAYMONGO_CONFIG.publicKey)}`,
      },
      body: JSON.stringify(payload),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.errors?.[0]?.detail || 'Failed to create payment method')
    }

    const data = await response.json()
    return data.data
  } catch (error: any) {
    throw new Error(error.message || 'Payment processing failed')
  }
}

/**
 * Create a PayMongo Payment Intent
 * Now calls our backend API which securely handles the secret key
 */
export async function createPaymentIntent(amount: number, description: string): Promise<PayMongoPaymentIntent> {
  // PayMongo expects amount in centavos (multiply by 100)
  const amountInCentavos = Math.round(amount * 100)

  try {
    // Call our backend API instead of PayMongo directly
    const response = await fetch(`${PAYMONGO_CONFIG.apiUrl}/api/payment/create-intent`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      credentials: 'include',
      body: JSON.stringify({
        amount: amountInCentavos,
        description: description,
      }),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || error.error || 'Failed to create payment intent')
    }

    const data = await response.json()
    return {
      id: data.payment_intent_id || data.client_key,
      attributes: {
        amount: amountInCentavos,
        currency: 'PHP',
        status: 'awaiting_payment_method',
        payment_method_allowed: ['card'],
      }
    }
  } catch (error: any) {
    throw new Error(error.message || 'Payment intent creation failed')
  }
}

/**
 * Attach Payment Method to Payment Intent
 * Now calls our backend API which securely handles the secret key
 */
export async function attachPaymentIntent(
  paymentIntentId: string,
  paymentMethodId: string
): Promise<PayMongoPaymentIntent> {
  try {
    const response = await fetch(`${PAYMONGO_CONFIG.apiUrl}/api/payment/attach-intent`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      credentials: 'include',
      body: JSON.stringify({
        payment_intent_id: paymentIntentId,
        payment_method_id: paymentMethodId,
        return_url: `${window.location.origin}/order-confirmed`,
      }),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || error.error || 'Failed to attach payment method')
    }

    const data = await response.json()
    return {
      id: data.payment_intent_id,
      attributes: {
        amount: data.amount || 0,
        currency: 'PHP',
        status: data.status || 'processing',
        payment_method_allowed: ['card'],
      }
    }
  } catch (error: any) {
    throw new Error(error.message || 'Payment attachment failed')
  }
}

/**
 * Process complete card payment
 */
export async function processCardPayment(
  cardDetails: CardDetails,
  amount: number,
  description: string
): Promise<{ success: boolean; paymentIntentId: string; status: string }> {
  try {
    // Step 1: Create payment method (tokenize card)
    const paymentMethod = await createPaymentMethod(cardDetails)

    // Step 2: Create payment intent
    const paymentIntent = await createPaymentIntent(amount, description)

    // Step 3: Attach payment method to intent
    const result = await attachPaymentIntent(paymentIntent.id, paymentMethod.id)

    return {
      success: result.attributes.status === 'succeeded' || result.attributes.status === 'processing',
      paymentIntentId: result.id,
      status: result.attributes.status,
    }
  } catch (error: any) {
    throw error
  }
}

/**
 * Validate card number (Luhn algorithm)
 */
export function validateCardNumber(cardNumber: string): boolean {
  const cleaned = cardNumber.replace(/\s/g, '')
  if (!/^\d{13,19}$/.test(cleaned)) return false

  let sum = 0
  let isEven = false

  for (let i = cleaned.length - 1; i >= 0; i--) {
    let digit = parseInt(cleaned[i])

    if (isEven) {
      digit *= 2
      if (digit > 9) digit -= 9
    }

    sum += digit
    isEven = !isEven
  }

  return sum % 10 === 0
}

/**
 * Format card number with spaces
 */
export function formatCardNumber(value: string): string {
  const cleaned = value.replace(/\s/g, '')
  const chunks = cleaned.match(/.{1,4}/g) || []
  return chunks.join(' ').substring(0, 19) // Max 16 digits + 3 spaces
}

/**
 * Get card type from number
 */
export function getCardType(cardNumber: string): string {
  const cleaned = cardNumber.replace(/\s/g, '')

  if (/^4/.test(cleaned)) return 'visa'
  if (/^5[1-5]/.test(cleaned)) return 'mastercard'
  if (/^3[47]/.test(cleaned)) return 'amex'
  if (/^6(?:011|5)/.test(cleaned)) return 'discover'
  if (/^35/.test(cleaned)) return 'jcb'

  return 'unknown'
}
